helix-lock/src/main/java/org/apache/helix/lock/helix/ZKDistributedNonblockingLock.java

```diff
@@ -125,6 +211,79 @@ public void close() {
     _baseDataAccessor.close();
   }
 
+  @Override
+  public void handleDataChange(String dataPath, Object data) throws Exception {
+    Stat stat = new Stat();
+    ZNRecord readData =
+        _baseDataAccessor.get(dataPath, stat, AccessOption.THROW_EXCEPTION_IFNOTEXIST);
+    LockInfo lockInfo = new LockInfo(readData);
+    // We are the current owner
+    if (lockInfo.getOwner().equals(_userId)) {
+      if (lockInfo.getRequestorId() == null || lockInfo.getPriority() > lockInfo
+          .getRequestorPriority()) {
+        LOG.info("We do not need to handle this data change");
+      } else {
+        _lockListener.onCleanupNotification();
```

Comment:
Not about code, but this means we don't have control on when the cleanup starts.
Would it be possible that the cleaner lock holder also updates the node so we are sure that cleaning up has been started?

Reply:
Clients need to implement the cleanup callback in their logic. We only trigger the registered listener.

Reply:
My proposal is that do the following steps in our code.

1. update znode about the cleanup start time.
2. trigger clean up callbacks.

So the cleanup start time is a more accurate sign than guessing.
I'm fine if we just add a TODO here about step 1.

Reply:
I agree with you that current time calculation is not accurate enough. But even with recording cleanup start time in znode, we still have a problem letting the requestor know the start time and adjust its waiting. Right now, the waiting time is calculated when requestor updates the znode, and cleanup update znode happens after requestor updates znode. How can we adjust the waiting time for requestor in this case?

Reply:
The proposal is that the waiting time should start once cleanup starts. But that could be too complicated. So let's just add a TODO here. We shall only do this when it is proved to be required.